src/sys_task.c

```diff
@@ -0,0 +1,165 @@
+#include "global.h"
+#include "sys_task.h"
+
+void SysTaskQueue_InitTask(SysTaskQueue *queue, SysTask *task);
+void SysTaskQueue_InitStack(SysTaskQueue *queue);
+SysTask *SysTaskQueue_CreateTask(SysTaskQueue *queue);
+BOOL SysTaskQueue_DeleteTask(SysTaskQueue *queue, SysTask *task);
+void SysTaskQueue_Init(SysTaskQueue *queue);
+SysTask *SysTaskQueue_InsertTaskCore(SysTaskQueue *queue, SysTaskFunc func, void *data, u32 priority);
+
+void SysTaskQueue_InitTask(SysTaskQueue *queue, SysTask *task) {
+    task->queue = queue;
+    task->prev = task->next = &queue->headSentinel;
+    task->priority = 0;
+    task->data = NULL;
+    task->func = NULL;
+}
+
+void SysTaskQueue_InitStack(SysTaskQueue *queue) {
+    for (int i = 0; i < queue->limit; ++i) {
+        SysTaskQueue_InitTask(queue, &queue->taskList[i]);
+        queue->taskStack[i] = &queue->taskList[i];
+    }
+    queue->activeCount = 0;
+}
+
+SysTask *SysTaskQueue_CreateTask(SysTaskQueue *queue) {
+    if (queue->activeCount == queue->limit) {
+        return NULL;
+    }
+    SysTask *ret = queue->taskStack[queue->activeCount];
+    ++queue->activeCount;
+    return ret;
+}
+
+BOOL SysTaskQueue_DeleteTask(SysTaskQueue *queue, SysTask *task) {
+    if (queue->activeCount == 0) {
+        return FALSE;
+    }
+    task->queue = queue;
+    task->prev = task->next = &queue->headSentinel;
+    task->priority = 0;
+    task->data = NULL;
+    task->func = NULL;
+    --queue->activeCount;
+    queue->taskStack[queue->activeCount] = task;
+    return TRUE;
+}
+
+u32 SysTaskQueue_GetArenaSize(u32 num) {
+    return num * (sizeof(SysTask) + sizeof(SysTask *)) + sizeof(SysTaskQueue);
+}
+
+SysTaskQueue *SysTaskQueue_PlacementNew(u32 num, void *p_mem) {
+    GF_ASSERT(p_mem != NULL);
+
+    SysTaskQueue *ret = (SysTaskQueue *)p_mem;
+    SysTask **taskStack = (SysTask **)((char *)p_mem + sizeof(SysTaskQueue));
+    ret->taskStack = taskStack;
+    SysTask *taskList = (SysTask *)((char *)taskStack + num * sizeof(SysTask *));
```

Comment:
can you rewrite this in terms of arrays ie taskStack[num]

Reply:
well whaddaya know, it still matches this way